GUROBI: Update Fragile Tests for new `gurobipy` version

## Summary

With the release of `gurobipy` version 11.0.0, some of the more fragile tests are failing. To enable us to cut the 6.7.0 release, we are pinning the version to 10.0.3. This issue serves as a reminder to update the fragile tests for the new and improved `gurobipy`.

### Rationale

We want to stay updated on the newest versions of solvers.

### Description

```
 =========================== short test summary info ============================
FAILED pyomo/contrib/appsi/solvers/tests/test_gurobi_persistent.py::TestGurobiPersistent::test_nonconvex_qcp_objective_bound_1 - AssertionError: -4.0 != None
FAILED pyomo/contrib/appsi/solvers/tests/test_gurobi_persistent.py::TestGurobiPersistent::test_nonconvex_qcp_objective_bound_2 - AssertionError: -4.0 != -6 within 7 places (2.0 difference)
==== 2 failed, 12400 passed, 5028 skipped, 29 xfailed in 737.77s (0:12:17) =====
```

These are the two fragile tests.



<!-- If you want to ask a question (how to use Pyomo, what it can do, etc.), please refer to our [online documentation](https://github.com/Pyomo/pyomo#getting-help) on ways to get help. -->

--- a/.github/workflows/test_branches.yml
+++ b/.github/workflows/test_branches.yml
@@ -286,7 +286,7 @@ jobs:
         if test -z "${{matrix.slim}}"; then
             python -m pip install --cache-dir cache/pip cplex docplex \
                 || echo "WARNING: CPLEX Community Edition is not available"
-            python -m pip install --cache-dir cache/pip gurobipy==10.0.3 \
+            python -m pip install --cache-dir cache/pip gurobipy \
                 || echo "WARNING: Gurobi is not available"
             python -m pip install --cache-dir cache/pip xpress \
                 || echo "WARNING: Xpress Community Edition is not available"
@@ -369,7 +369,7 @@ jobs:
         if test -z "${{matrix.slim}}"; then
             PYVER=$(echo "py${{matrix.python}}" | sed 's/\.//g')
             echo "Installing for $PYVER"
-            for PKG in 'cplex>=12.10' docplex 'gurobi=10.0.3' xpress cyipopt pymumps scip; do
+            for PKG in 'cplex>=12.10' docplex gurobi xpress cyipopt pymumps scip; do
                 echo ""
                 echo "*** Install $PKG ***"
                 # conda can literally take an hour to determine that a

--- a/.github/workflows/test_pr_and_main.yml
+++ b/.github/workflows/test_pr_and_main.yml
@@ -309,7 +309,7 @@ jobs:
         if test -z "${{matrix.slim}}"; then
             python -m pip install --cache-dir cache/pip cplex docplex \
                 || echo "WARNING: CPLEX Community Edition is not available"
-            python -m pip install --cache-dir cache/pip gurobipy==10.0.3 \
+            python -m pip install --cache-dir cache/pip gurobipy \
                 || echo "WARNING: Gurobi is not available"
             python -m pip install --cache-dir cache/pip xpress \
                 || echo "WARNING: Xpress Community Edition is not available"
@@ -392,7 +392,7 @@ jobs:
         if test -z "${{matrix.slim}}"; then
             PYVER=$(echo "py${{matrix.python}}" | sed 's/\.//g')
             echo "Installing for $PYVER"
-            for PKG in 'cplex>=12.10' docplex 'gurobi=10.0.3' xpress cyipopt pymumps scip; do
+            for PKG in 'cplex>=12.10' docplex gurobi xpress cyipopt pymumps scip; do
                 echo ""
                 echo "*** Install $PKG ***"
                 # conda can literally take an hour to determine that a

--- a/pyomo/contrib/appsi/solvers/tests/test_gurobi_persistent.py
+++ b/pyomo/contrib/appsi/solvers/tests/test_gurobi_persistent.py
@@ -213,9 +213,9 @@ class TestGurobiPersistent(unittest.TestCase):
         opt.config.load_solution = False
         res = opt.solve(m)
         if opt.version() < (11, 0):
-            self.assertEqual(res.incumbent_objective, None)
+            self.assertEqual(res.best_feasible_objective, None)
         else:
-            self.assertEqual(res.incumbent_objective, -4)
+            self.assertEqual(res.best_feasible_objective, -4)
         self.assertAlmostEqual(res.best_objective_bound, -8)
 
     def test_nonconvex_qcp_objective_bound_2(self):

--- a/pyomo/contrib/appsi/solvers/tests/test_gurobi_persistent.py
+++ b/pyomo/contrib/appsi/solvers/tests/test_gurobi_persistent.py
@@ -240,9 +240,9 @@ class TestGurobiPersistent(unittest.TestCase):
         res = opt.solve(m)
         self.assertAlmostEqual(res.best_feasible_objective, -4)
         if opt.version() < (11, 0):
-            self.assertAlmostEqual(res.objective_bound, -6)
+            self.assertAlmostEqual(res.best_objective_bound, -6)
         else:
-            self.assertAlmostEqual(res.objective_bound, -4)
+            self.assertAlmostEqual(res.best_objective_bound, -4)
 
     def test_range_constraints(self):
         m = pe.ConcreteModel()

--- a/pyomo/contrib/appsi/solvers/tests/test_gurobi_persistent.py
+++ b/pyomo/contrib/appsi/solvers/tests/test_gurobi_persistent.py
@@ -191,12 +191,16 @@ class TestGurobiPersistentSimpleLPUpdates(unittest.TestCase):
 
 class TestGurobiPersistent(unittest.TestCase):
     def test_nonconvex_qcp_objective_bound_1(self):
-        # the goal of this test is to ensure we can get an objective bound
-        # for nonconvex but continuous problems even if a feasible solution
-        # is not found
+        # the goal of this test is to ensure we can get an objective
+        # bound for nonconvex but continuous problems even if a feasible
+        # solution is not found
         #
-        # This is a fragile test because it could fail if Gurobi's algorithms improve
-        # (e.g., a heuristic solution is found before an objective bound of -8 is reached
+        # This is a fragile test because it could fail if Gurobi's
+        # algorithms improve (e.g., a heuristic solution is found before
+        # an objective bound of -8 is reached
+        #
+        # Update: as of Gurobi 11, this test no longer tests the
+        # intended behavior (the solver has improved)
         m = pe.ConcreteModel()
         m.x = pe.Var(bounds=(-5, 5))
         m.y = pe.Var(bounds=(-5, 5))
@@ -208,14 +212,22 @@ class TestGurobiPersistent(unittest.TestCase):
         opt.gurobi_options['BestBdStop'] = -8
         opt.config.load_solution = False
         res = opt.solve(m)
-        self.assertEqual(res.best_feasible_objective, None)
+        if opt.version() < (11, 0):
+            self.assertEqual(res.incumbent_objective, None)
+        else:
+            self.assertEqual(res.incumbent_objective, -4)
         self.assertAlmostEqual(res.best_objective_bound, -8)
 
     def test_nonconvex_qcp_objective_bound_2(self):
-        # the goal of this test is to ensure we can best_objective_bound properly
-        # for nonconvex but continuous problems when the solver terminates with a nonzero gap
+        # the goal of this test is to ensure we can best_objective_bound
+        # properly for nonconvex but continuous problems when the solver
+        # terminates with a nonzero gap
         #
-        # This is a fragile test because it could fail if Gurobi's algorithms change
+        # This is a fragile test because it could fail if Gurobi's
+        # algorithms change
+        #
+        # Update: as of Gurobi 11, this test no longer tests the
+        # intended behavior (the solver has improved)
         m = pe.ConcreteModel()
         m.x = pe.Var(bounds=(-5, 5))
         m.y = pe.Var(bounds=(-5, 5))
@@ -227,7 +239,10 @@ class TestGurobiPersistent(unittest.TestCase):
         opt.gurobi_options['MIPGap'] = 0.5
         res = opt.solve(m)
         self.assertAlmostEqual(res.best_feasible_objective, -4)
-        self.assertAlmostEqual(res.best_objective_bound, -6)
+        if opt.version() < (11, 0):
+            self.assertAlmostEqual(res.objective_bound, -6)
+        else:
+            self.assertAlmostEqual(res.objective_bound, -4)
 
     def test_range_constraints(self):
         m = pe.ConcreteModel()

--- a/pyomo/contrib/solver/tests/solvers/test_gurobi_persistent.py
+++ b/pyomo/contrib/solver/tests/solvers/test_gurobi_persistent.py
@@ -192,8 +192,12 @@ class TestGurobiPersistent(unittest.TestCase):
         # for nonconvex but continuous problems even if a feasible solution
         # is not found
         #
-        # This is a fragile test because it could fail if Gurobi's algorithms improve
-        # (e.g., a heuristic solution is found before an objective bound of -8 is reached
+        # This is a fragile test because it could fail if Gurobi's
+        # algorithms improve (e.g., a heuristic solution is found before
+        # an objective bound of -8 is reached
+        #
+        # Update: as of Gurobi 11, this test no longer tests the
+        # intended behavior (the solver has improved)
         m = pe.ConcreteModel()
         m.x = pe.Var(bounds=(-5, 5))
         m.y = pe.Var(bounds=(-5, 5))
@@ -206,14 +210,22 @@ class TestGurobiPersistent(unittest.TestCase):
         opt.config.load_solutions = False
         opt.config.raise_exception_on_nonoptimal_result = False
         res = opt.solve(m)
-        self.assertEqual(res.incumbent_objective, None)
+        if opt.version() < (11, 0):
+            self.assertEqual(res.incumbent_objective, None)
+        else:
+            self.assertEqual(res.incumbent_objective, -4)
         self.assertAlmostEqual(res.objective_bound, -8)
 
     def test_nonconvex_qcp_objective_bound_2(self):
-        # the goal of this test is to ensure we can objective_bound properly
-        # for nonconvex but continuous problems when the solver terminates with a nonzero gap
+        # the goal of this test is to ensure we can objective_bound
+        # properly for nonconvex but continuous problems when the solver
+        # terminates with a nonzero gap
         #
-        # This is a fragile test because it could fail if Gurobi's algorithms change
+        # This is a fragile test because it could fail if Gurobi's
+        # algorithms change
+        #
+        # Update: as of Gurobi 11, this test no longer tests the
+        # intended behavior (the solver has improved)
         m = pe.ConcreteModel()
         m.x = pe.Var(bounds=(-5, 5))
         m.y = pe.Var(bounds=(-5, 5))
@@ -225,7 +237,10 @@ class TestGurobiPersistent(unittest.TestCase):
         opt.config.solver_options['MIPGap'] = 0.5
         res = opt.solve(m)
         self.assertAlmostEqual(res.incumbent_objective, -4)
-        self.assertAlmostEqual(res.objective_bound, -6)
+        if opt.version() < (11, 0):
+            self.assertAlmostEqual(res.objective_bound, -6)
+        else:
+            self.assertAlmostEqual(res.objective_bound, -4)
 
     def test_range_constraints(self):
         m = pe.ConcreteModel()

